File: 03_Lifecycles/04_SavedState/app/src/main/java/com/codeingnight/android/scoreboard/MainActivity.java
package com.codeingnight.android.scoreboard;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.SavedStateViewModelFactory;
import androidx.lifecycle.ViewModelProvider;

import android.os.Bundle;

import com.codeingnight.android.scoreboard.databinding.ActivityMainBinding;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = "MainActivity";
    ScoreboardViewModel viewModel;
    ActivityMainBinding binding;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        viewModel = new ViewModelProvider(this, new SavedStateViewModelFactory(getApplication(), this)).get(ScoreboardViewModel.class);
        binding.setViewModel(viewModel);
        binding.setLifecycleOwner(this);
    }
}

File: 01_AndroidUI/01_CommonUIWidgets/03_Chronometer/src/main/java/com/codingnight/example/chronometer/MainActivity.kt
package com.codingnight.example.chronometer

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Chronometer
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import com.codingnight.example.chronometer.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {

    private lateinit var meter: Chronometer
    lateinit var viewModel: MeterViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding: ActivityMainBinding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        meter = binding.meter
        viewModel = ViewModelProvider(this).get(MeterViewModel::class.java)
        binding.viewModel = viewModel
        binding.lifecycleOwner = this
    }

    override fun onPause() {
        super.onPause()
        viewModel.setElapsedTime(meter.base)
        meter.stop()
    }

    override fun onResume() {
        super.onResume()
        meter.base = viewModel.getBase()
        meter.start()
    }
}